pkg/gadgets/process-collector/gadget.go

```diff
@@ -62,22 +66,41 @@ func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
 }
 
 func (t *Trace) Collect(trace *gadgetv1alpha1.Trace) {
-	selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
-	if len(t.resolver.GetContainersBySelector(selector)) == 0 {
+	events, err := tracer.RunCollector(gadgets.TracePinPath(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name))
+	if err != nil {
+		gadgets.CleanupTraceStatus(trace)
+		trace.Status.OperationError = err.Error()
+		return
+	}
+
+	if len(events) == 0 {
 		gadgets.CleanupTraceStatus(trace)
 		trace.Status.OperationWarning = "No container matches the requested filter"
 		return
 	}
 
-	output, err := tracer.RunCollector(
-		gadgets.TracePinPath(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name),
-	)
+	for i, event := range events {
+		c := t.resolver.LookupContainerByMntns(event.MntNsId)
+		if c == nil {
+			trace.Status.OperationWarning += fmt.Sprintf("No container found for mntsnsid %v\n", event.MntNsId)
+			continue
+		}
+
+		events[i].Event = eventtypes.Event{
+			Namespace: c.Namespace,
+			Pod:       c.Podname,
+			Container: c.Name,
+		}
+	}
```

Comment:
In https://github.com/kinvolk/inspektor-gadget/pull/438 and https://github.com/kinvolk/inspektor-gadget/pull/425 I followed a slightly different approach. Instead of having this in the gadget.go file I put this logic in the tracer. My reasoning is that the tracer should return the complete events and the gadget shouldn't worry about filling them. It would make the tracer package reusable without having to add logic outside it. 

What I did exactly was to pass the resolver interface to `NewTracer()` (`RunCollector()` in this case), and then that function would use it to fill the event. The logic in this file will only be to run the collector and marshall the result. 

What do you think about this approach? Are you aware of any advantages / disadvantage?

Reply:
I agree with you this is a bit odd to have part of the event filled in the tracer and the other in `gadget.go`.
Also, I would say it seems better to handle everything as near of eBPF code as possible.
I will try to implement the same reasoning here.